### PR TITLE
Add normalizes_type to apply common normalizations

### PR DESCRIPTION
**Pull request by codergeek121**
### Motivation / Background

Many apps use the https://github.com/rmm5t/strip_attributes gem to strip most string attributes.
The new `normalizes` API is a perfect fit for this. One "downside" in contrast to the strip_attributes gem, is that you have to explicitly list all attributes you want to normalize. It would be nice to have a simple way to e.g. "strip all string attributes in most models".

### Checklist

Before submitting the PR make sure the following are checked:

* [x] This Pull Request is related to one change. Changes that are unrelated should be opened in separate PRs.
* [x] Commit message has a detailed description of what changed and why. If this PR fixes a related issue include it in the commit message. Ex: `[Fix #issue-number]`
* [x] Tests are added or updated if you fix a bug or add a feature.
* [x] CHANGELOG files are updated for the changed libraries if there is a behavior change or additional feature. Minor bug fixes and documentation changes should not be included.

### CODE REVIEW

**Comment by rubyhcm** [comment]
OK

**Comment by rubyhcm** [comment]
Good PR bro

File: activerecord/CHANGELOG.md [hunk 1]
```diff
@@ -1,2 +1,11 @@
+*   Add `normalizes_type` to apply normalizations to multiple attributes of the same type.
+
+    ```ruby
+    class User < ActiveRecord::Base
+      normalizes_type :string, except: :name, with: -> attribute { attribute.strip }
```
Comment (casperisfine) [inline activerecord/CHANGELOG.md:5]: I like the capability but I'm not sure about the API in it's current state. 

to be thought about: cc @byroot

**Comment by rafaelfranca** [comment]
I don't like APIs that expect people to add exception lists. If you want to normalize attributes, why you don't explicitly don't tell which one?

This also forces the framework to loads the schema when loading the model instead when it is really necessary.

**Comment by jonathanhefner** [comment]
`normalizes` accepts a list of attributes, so I think that gets us halfway there.  For the remaining half, we could add a method that returns the list of attributes we're interested in.  It could look like:

  ```ruby
  class Post < ActiveRecord::Base
    normalizes attributes_of_type(:string), with: -> { _1.strip }
  end
  ```

And if you need to exclude attributes, you could use [`excluding`](https://api.rubyonrails.org/v7.1/classes/Array.html#method-i-excluding):

  ```ruby
  class User < ActiveRecord::Base
    normalizes attributes_of_type(:string).excluding(:name), with: -> { _1.strip }
  end
  ```

To that end, I've submitted #49519.

**Comment by codergeek121** [comment]
Thank you @jonathanhefner for your PR 👍 
I think it will not be easily possible to avoid loading the schema too early without adding some kind of API to `ActiveRecord::Normalization`.
But imo your PR still makes sense, maybe just without the reference to normalization. Imo `ModelSchema::ClassMethods#attributes_of_type` would be super useful in this PR to have.

I added https://github.com/rails/rails/pull/49314/commits/63996c2177610bd15a515510728d07563cbfee68 which removes the `except:` option and tries to create the normalizations after the schema is loaded by hooking into `load_schema!`. It's still buggy (and can't figure out the issue currently), but I decided to push if you want to have a look as well.

**Comment by jonathanhefner** [comment]
@codergeek121 Thank you for continuing to work on this!  Indeed, a different approach than #49519 is needed.  I have been working a PR and will submit it soon.  Unfortunately, there are a few edge cases that a `load_schema!`-based approach cannot address.  But I will be sure to add you to the PR as a co-author!

**Comment by willianveiga** [comment]
Should we close this PR in favor of #49646?

**Pull request closed without merging**